### PR TITLE
RzArch: Support for memory mapping the ROM sizes

### PR DESCRIPTION
**Pull request by officialcjunior**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Supports adding ROM related data as `iS` mappings. 

The value of `ROM_ADDRESS` and `ROM_SIZE` is parsed from the CPU Profile. During `aaa`, `rz_analysis_add_io_registers_map` is called which creates the new `RzBinSection`.

**Test plan**

Current state:
```
$ ./build/binrz/rizin/rizin -e asm.cpu=ATmega16 test/bins/firmware/arduino_avr.bin
 -- Use 'rz-bin -ris' to get the import/export symbols of any binary.
[0x00000158]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
[x] Analyze function calls (aac)
(...)
[0x00000158]> iS~.rom
 8   0x00001fff  0x4000 0x00001fff  0x4000 -r-x .rom      NULL     
```

- [x] Handle the case where a map already exists.

**Closing issues**

None

### CODE REVIEW

File: librz/core/canalysis.c
Comment (wargio) [inline librz/core/canalysis.c]: i would say here, that `if (file_offset >= rom_address && file_offset < (rom_address + rom_size))` then throw that warning.

File: librz/core/canalysis.c
Comment (wargio) [inline librz/core/canalysis.c]: `RZ_LOG_WARN("Cannot map the ROM section at %" PFMT64x " with a size of %" PFMT64x " due file being mapped at 0x%" PFMT64x "\n", rom_address, rom_size, file_offset);`

File: librz/core/cmd_analysis.c
Comment (wargio) [inline librz/core/cmd_analysis.c]: move this within `rz_core_analysis_everything`

**Review by wargio — CHANGES_REQUESTED** [review]
And add a test to test the warning. keep in mind since this is a warning, you should verify the stderr via `EXPECT_ERR` in the test.

File: librz/core/cmd_analysis.c
Comment (officialcjunior) [reply librz/core/cmd_analysis.c:8008]: I had done that in https://github.com/rizinorg/rizin/pull/1228/commits/227f9c3a606430469615d433359e6952edbcca70#diff-e4de03bacabad8d61bf51daee587cbf69970dae65a272486119cdaebfbf9e87aR7174, but from what I had seen, calling it from `rz_core_analysis_everything` broke a lot of analysis stuff and tests and I am still trying to find a reason.

File: test/db/analysis/avr
Comment (wargio) [inline test/db/analysis/avr]: Use the `REGEXP_FILTER_ERR` to add a regex for the warning, so everything else is ignored.
See https://github.com/rizinorg/rizin/tree/dev/test#commands-tests

Comment (officialcjunior) [reply test/db/analysis/avr]: Can you name a method/idea to match it?
I had tried that for things I could think of like `^WARNING` and all which didn't seem to work

Comment (wargio) [reply test/db/analysis/avr]: can you write here the full line?

File: librz/core/canalysis.c
Comment (wargio) [inline librz/core/canalysis.c]: in the original code this was at the beginning, before the analysis. does this change something in the analysis loop?

File: test/db/analysis/avr
Comment (officialcjunior) [reply test/db/analysis/avr]: So, when I put the whole line:
```
REGEXP_FILTER_ERR=(WARNING: Cannot map the ROM section at 1fff with a size of 4000 due file being mapped at 0x1fff)
EXPECT_ERR=<<EOF
WARNING: Cannot map the ROM section at 1fff with a size of 4000 due file being mapped at 0x1fff
EOF
RUN
```
It says it finds a mismatch.

```

#####################
 1 failed test(s) 🚨
#####################

RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -e asm.cpu=ATmega16 -m 8191 -Qc 'aaa
om
' bins/firmware/arduino_avr.bin
-- stderr
--- expected
+++ actual
@@ -1,1 +1,0 @@
-WARNING: Cannot map the ROM section at 1fff with a size of 4000 due file being mapped at 0x1fff

Wat do?    (f)ix ✅️️️    (i)gnore 🙈    (b)roken ☠️️️    (c)ommands ⌨️    (q)uit 🚪
> 

```
And when I press `f` and fix it, the whole stderr is written back on the file again. And when I try to run rz-test again, this is what I get:

```
#####################
 1 failed test(s) 🚨
#####################

RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -e asm.cpu=ATmega16 -m 8191 -Qc 'aaa
om
' bins/firmware/arduino_avr.bin
-- stderr
--- expected
+++ actual
@@ -1,15 +1,0 @@
-Warning: using oba to load the syminfo from different mapaddress.
[x] Analyze all flags starting with sym. and entry0 (aa))
[x] Analyze function calls (aac)
[x] find and analyze function preludes (aap))
[x] Analyze len bytes of instructions for references (aar))
[x] Check for classess
[x] Finding xrefs in noncode section with analysis.in=io.mapss
[x] Analyze value pointers (aav))
[x] Value from 0x00000000 to 0x00003722 (aav))
[x] 0x00000000-0x00003722 in 0x0-0x3722 (aav))
[x] Emulate functions to find computed references (aaef))
[x] Type matching analysis for all functions (aaft))
[x] Propagate noreturn informationn
[x] Use -AA or aaaa to perform additional experimental analysis..
-WARNING: Cannot map the ROM section at 1fff with a size of 4000 due file being mapped at 0x1fff

Wat do?    (f)ix ✅️️️    (i)gnore 🙈    (b)roken ☠️️️    (c)ommands ⌨️    (q)uit 🚪
> f

``` 

and this never gets fixed when I press `f`.

Comment (XVilka) [reply test/db/analysis/avr]: I think the autofixing feature wasn't updated for handling regexps. So this is expected somehow.

File: librz/core/canalysis.c
Comment (wargio) [inline librz/core/canalysis.c]: add `{ ... }`

File: test/db/analysis/avr
Comment (wargio) [reply test/db/analysis/avr]: it's ok. we found another way to do this.

**Comment by officialcjunior** [comment]
There's one single test that is broken.

**Review by wargio — APPROVED** [review]

File: librz/core/canalysis.c
Comment (XVilka) [inline librz/core/canalysis.c]: If you add a new API function - don't forget Doxygen please.

Comment (XVilka) [reply librz/core/canalysis.c]: Also please return `false` in case of either `o` and `analysis` are NULL. Also please add `rz_return_val_if_fail(o && analysis, false)`;.

**Review by XVilka — APPROVED** [review]
Overall is good, just please address these minor suggestions. Also lets wait for one more approval from either @ret2libc or @thestr4ng3r

File: librz/core/canalysis.c
Comment (XVilka) [inline librz/core/canalysis.c]: This is not a brief. Please split it into one-line `brief` and more detailed explanation below

File: librz/core/canalysis.c
Comment (XVilka) [inline librz/core/canalysis.c]: Wrong indentation.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Pull request merged**
